source/custom/nixnet_service.custom.cpp

```diff
@@ -1,2 +1,207 @@
+    }
+    try {
+        auto session_ref_grpc_session = request->session_ref();
+        nxSessionRef_t session_ref = session_repository_->access_session(session_ref_grpc_session.id(), session_ref_grpc_session.name());
+        u32 state_id;
+        switch (request->state_id_enum_case()) {
+            case nixnet_grpc::ReadStateRequest::StateIdEnumCase::kStateId: {
+                state_id = static_cast<u32>(request->state_id());
+                break;
+            }
+            case nixnet_grpc::ReadStateRequest::StateIdEnumCase::kStateIdRaw: {
+                state_id = static_cast<u32>(request->state_id_raw());
+                break;
+            }
+            case nixnet_grpc::ReadStateRequest::StateIdEnumCase::STATE_ID_ENUM_NOT_SET: {
+                return ::grpc::Status(::grpc::INVALID_ARGUMENT, "The value for state_id was not specified or out of range");
+                break;
+            }
+        }
+
+        u32 state_size = request->state_size();
+        u64 time_stamp_100ns;
+        u32 comm_state;
+        FlexRayStats flex_ray_stats;
+        J1939CommState J1939_comm_state;
+        TimeLocalNetwork time_local_network;
+        void* state_value;
+
+        switch(state_id){
+            case nixnet_grpc::ReadState::READ_STATE_STATE_TIME_CURRENT : 
+            case nixnet_grpc::ReadState::READ_STATE_STATE_TIME_COMMUNICATING : 
+            case nixnet_grpc::ReadState::READ_STATE_STATE_TIME_START :{
+                state_value = &time_stamp_100ns;
+                break;
+            }
+            case nixnet_grpc::ReadState::READ_STATE_STATE_CAN_COMM :
+            case nixnet_grpc::ReadState::READ_STATE_STATE_FLEX_RAY_COMM :
+            case nixnet_grpc::ReadState::READ_STATE_STATE_LIN_COMM :
+            case nixnet_grpc::ReadState::READ_STATE_STATE_SESSION_INFO :{
+                state_value = &comm_state;
+                break;
+            }
+            case nixnet_grpc::ReadState::READ_STATE_STATE_FLEX_RAY_STATS :{
+                state_value = &flex_ray_stats;
+                break;
+            }
+            case nixnet_grpc::ReadState::READ_STATE_STATE_J1939_COMM :{
+                state_value = &J1939_comm_state;
+                break;
+            }
+            case nixnet_grpc::ReadState::READ_STATE_STATE_TIME_CURRENT_2 : 
+            case nixnet_grpc::ReadState::READ_STATE_STATE_TIME_COMMUNICATING_2 : 
+            case nixnet_grpc::ReadState::READ_STATE_STATE_TIME_START_2 :{
+                state_value = &time_local_network;
+                break;
+            }
+        }
+        nxStatus_t fault {};
+        auto status = library_->ReadState(session_ref, state_id, state_size, state_value, &fault);
+        response->set_status(status);
+        if (status == 0) {
+            switch(state_id){
+            case nixnet_grpc::ReadState::READ_STATE_STATE_TIME_CURRENT :{
+                response->mutable_state_value()->set_time_current(*((u64*)state_value));
```

Comment:
Couldn't all of these reference the stack field and avoid the type cast? So, in this case just pass in `time_stamp_100ns`?

Reply:
Yes, done